File: content/snippets/_harness/fake-model.mjs
/**
 * Local doubles for self-hosted specialised models (rung N2).
 * JavaScript counterpart of _harness/fake_model.py.
 */

export class FakeClassifier {
  constructor(scores, defaultScores = {}) {
    this.scores = scores;
    this.default = defaultScores;
    this.calls = [];
  }

  async predict(texts) {
    this.calls.push([...texts]);
    return texts.map((t) => this.scores[t] ?? this.default);
  }
}

export class FakeEncoder {
  constructor(dimensions = 32) {
    this.dimensions = dimensions;
    this.calls = [];
  }

  async encode(texts) {
    this.calls.push([...texts]);
    return texts.map((t) => this.#vector(t));
  }

  #vector(text) {
    const vector = new Array(this.dimensions).fill(0);
    for (const word of words(text)) {
      vector[stableHash(word) % this.dimensions] += 1;
    }
    const norm = Math.hypot(...vector);
    return norm ? vector.map((v) => v / norm) : vector;
  }
}

export class FakeSeq2Seq {
  constructor(outputs, defaultOutput = '') {
    this.outputs = outputs;
    this.default = defaultOutput;
    this.calls = [];
  }

  async generate(text) {
    this.calls.push(text);
    return this.outputs[text] ?? this.default;
  }
}

export class FakeOCR {
  constructor(pages, confidence = 0.92) {
    this.pages = pages;
    this.confidence = confidence;
    this.calls = [];
  }

  async read(imagePath) {
    this.calls.push(imagePath);
    return { text: this.pages[imagePath] ?? '', confidence: this.confidence };
  }
}

/**
 * FNV-1a, 32 bits. Must agree exactly with its Python counterpart, so that a
 * snippet's two language versions score the same inputs identically.
 *
 * Math.imul is what makes that true: a plain `*` on numbers this large loses
 * precision past 2^53 and silently drifts from Python's exact integers.
 */
export function stableHash(word) {
  let h = 2166136261;
  for (const char of word) {
    h = Math.imul(h ^ char.codePointAt(0), 16777619) >>> 0;
  }
  return h;
}

function words(text) {
  return text
    .toLowerCase()
    .replace(/[^\p{L}\p{N}]+/gu, ' ')
    .split(' ')
    .filter(Boolean);
}

File: content/snippets/_harness/fake_model.py
"""
Local double for a self-hosted specialised model (rung N2).

Snippets on rung N2 load a real model: a distilled encoder, an embedding
model, an OCR engine, a translation model. Those weigh hundreds of megabytes
and cannot be downloaded on every build.

The snippet therefore takes its model as an injectable parameter with a real
default, so the reader sees the real loading code. The test injects one of the
doubles below.

Snippets tested this way are declared `verification: stubbed`.
"""

from __future__ import annotations

from typing import Any, Callable, Sequence


class FakeClassifier:
    """
    A classifier whose decisions are given, not learnt.

    Lets a test exercise thresholding, label mapping, batching and error
    handling in the snippet, without a real model.
    """

    def __init__(self, scores: dict[str, dict[str, float]], default: dict[str, float] | None = None):
        self.scores = scores
        self.default = default or {}
        self.calls: list[Sequence[str]] = []

    def predict(self, texts: Sequence[str]) -> list[dict[str, float]]:
        self.calls.append(list(texts))
        return [self.scores.get(t, self.default) for t in texts]


class FakeEncoder:
    """
    Deterministic sentence encoder.

    Same text gives the same vector; texts sharing words land closer together.
    Enough to exercise similarity, ranking and nearest-neighbour code paths.
    """

    def __init__(self, dimensions: int = 32):
        self.dimensions = dimensions
        self.calls: list[Sequence[str]] = []

    def encode(self, texts: Sequence[str]) -> list[list[float]]:
        self.calls.append(list(texts))
        return [self._vector(t) for t in texts]

    def _vector(self, text: str) -> list[float]:
        vector = [0.0] * self.dimensions
        for word in _words(text):
            vector[stable_hash(word) % self.dimensions] += 1.0
        norm = sum(v * v for v in vector) ** 0.5
        return [v / norm for v in vector] if norm else vector


class FakeSeq2Seq:
    """A translation or summarisation model whose outputs are given."""

    def __init__(self, outputs: dict[str, str], default: str = ""):
        self.outputs = outputs
        self.default = default
        self.calls: list[str] = []

    def generate(self, text: str, **_: Any) -> str:
        self.calls.append(text)
        return self.outputs.get(text, self.default)


class FakeOCR:
    """An OCR engine whose reading of each image is given."""

    def __init__(self, pages: dict[str, str], confidence: float = 0.92):
        self.pages = pages
        self.confidence = confidence
        self.calls: list[str] = []

    def read(self, image_path: str) -> dict[str, Any]:
        self.calls.append(image_path)
        return {"text": self.pages.get(image_path, ""), "confidence": self.confidence}


def stable_hash(word: str) -> int:
    """Stable across runs, unlike the built-in hash of a string."""
    h = 2166136261
    for char in word:
        h = ((h ^ ord(char)) * 16777619) & 0xFFFFFFFF
    return h


def _words(text: str) -> list[str]:
    return [w for w in "".join(c if c.isalnum() else " " for c in text.lower()).split() if w]


def loader_returning(model: Any) -> Callable[..., Any]:
    """Turns an object into a loader, for snippets that take a loader."""
    return lambda *_args, **_kwargs: model
